### PR TITLE
Set correct device ids in DDP [wip]

### LINKED ISSUE #3791: error=101 : invalid device ordinal
## 🐛 Bug

When the first entry of CUDA_VISIBLE_DEVICES > 0 and the `ddp` backend is used, a cuda error `invalid device ordinal` occurs.
After digging into the library's code I found the source of the issue in this function:
 https://github.com/PyTorchLightning/pytorch-lightning/blob/440f837f6d1b5fc44e6f04475fd2af20e2ed370d/pytorch_lightning/accelerators/ddp_backend.py#L151
which I copy and paste here:

```
def model_to_device(self, model, process_idx, is_master):
1    gpu_idx = process_idx

    # when using ddp, the master process (proc 0) continues running as the main one
    # this means that the local rank will always be 0
    # (even if cuda visible devices has other visible gpus)
    # this means that the master process needs to pull the 0th visible index as the device number
 2   if is_master:
 3       available_gpus = os.environ['CUDA_VISIBLE_DEVICES'].split(',')
 4       gpu_idx = int(available_gpus[self.trainer.local_rank])

 5   gpu_idx = int(os.environ.get('PL_DDP_PID', gpu_idx))

 6   self.trainer.root_gpu = gpu_idx
 7   torch.cuda.set_device(self.trainer.root_gpu)
 8   model.cuda(self.trainer.root_gpu)
```

Assume `is_master=True` , `self.trainer.local_rank=0` and `CUDA_VISIBLE_DEVICES=4,5,6,7` from the environment.
Then in line 4 `gpu_idx` becomes `4`. In line 5 `gpu_idx` remains unchanged because `PL_DDP_PID` is not defined.
Finally in line 7 we get the error because the device indexing has to be in the range [0,4). but we try to set the device to 4. 
The problem here is that gpu_idx is taken in absolute terms in line 4, while it should use a relative indexing in line 7.

### LINKED ISSUE #3865: NCCL error when using ddp with 2 gpus
## 🐛 Bug

I try to run pytorch lighting using ddp with 2 gpus. Running with one gpu works fine. Using fp16 vs not results in the same error. See the stacktrace at the end of the post to see the error. I also tried ddp2 and dp, but both of those fail with a different error.

### To Reproduce
Not sure. Let me know what I can do to diagnose.

I'm running my code on a cluster where each gpu is locked to one process. I'm using NCCL version 2.4.8. 

I tried pytorch-lightning versions `0.9.0`, `0.9.1rc4`, `0.10.0rc1`. All of them result in the same error. I'm running pytorch version `1.6`. 

### Expected behavior

I expected training to start running smoothly using both gpus.

### Environment
* CUDA:
        - GPU:
                - GeForce GTX 1080
                - GeForce GTX 1080
        - available:         True
        - version:           10.1
* Packages:
        - numpy:             1.18.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.6.0
        - pytorch-lightning: 0.10.0rc1
        - tqdm:              4.46.1
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - 
        - processor:         
        - python:            3.7.7
        - version:           #1 SMP Tue May 12 16:57:42 UTC 2020

### Additional context
Stacktrace and error.
```
initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/2
INFO:lightning:initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/2
LOCAL_RANK: 1 - CUDA_VISIBLE_DEVICES: [0,1]
INFO:lightning:LOCAL_RANK: 1 - CUDA_VISIBLE_DEVICES: [0,1]
Using native 16bit precision.
INFO:lightning:Using native 16bit precision.
initializing ddp: GLOBAL_RANK: 1, MEMBER: 2/2
INFO:lightning:initializing ddp: GLOBAL_RANK: 1, MEMBER: 2/2
lo-s4-039:21587:21587 [0] NCCL INFO Bootstrap : Using [0]fabric:10.204.67.89<0> [1]enp129s0f0:10.204.3.89<0>
lo-s4-039:21587:21587 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
libibverbs: Warning: couldn't open config directory '/etc/libibverbs.d'.
libibverbs: Warning: no userspace device-specific driver found for /sys/class/infiniband_verbs/uverbs1
libibverbs: Warning: no userspace device-specific driver found for /sys/class/infiniband_verbs/uverbs0
lo-s4-039:21587:21587 [0] NCCL INFO NET/IB : No device found.
lo-s4-039:21587:21587 [0] NCCL INFO NET/Socket : Using [0]fabric:10.204.67.89<0> [1]enp129s0f0:10.204.3.89<0>
NCCL version 2.4.8+cuda10.1
lo-s4-039:21614:21614 [1] NCCL INFO Bootstrap : Using [0]fabric:10.204.67.89<0> [1]enp129s0f0:10.204.3.89<0>
lo-s4-039:21614:21614 [1] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
libibverbs: Warning: couldn't open config directory '/etc/libibverbs.d'.
libibverbs: Warning: no userspace device-specific driver found for /sys/class/infiniband_verbs/uverbs1
libibverbs: Warning: no userspace device-specific driver found for /sys/class/infiniband_verbs/uverbs0
lo-s4-039:21614:21614 [1] NCCL INFO NET/IB : No device found.
lo-s4-039:21614:21614 [1] NCCL INFO NET/Socket : Using [0]fabric:10.204.67.89<0> [1]enp129s0f0:10.204.3.89<0>
lo-s4-039:21587:21646 [0] NCCL INFO Setting affinity for GPU 0 to 1fd001fd
lo-s4-039:21614:21647 [1] NCCL INFO Setting affinity for GPU 1 to 1fd001fd
lo-s4-039:21587:21646 [0] NCCL INFO Channel 00 :    0   1
lo-s4-039:21587:21646 [0] NCCL INFO Ring 00 : 0[1] -> 1[2] via P2P/IPC
lo-s4-039:21614:21647 [1] NCCL INFO Ring 00 : 1[2] -> 0[1] via P2P/IPC

lo-s4-039:21587:21646 [0] transport/p2p.cc:574 NCCL WARN failed to open CUDA IPC handle : 711 peer mapping resources exhausted
lo-s4-039:21587:21646 [0] NCCL INFO init.cc:669 -> 1
lo-s4-039:21587:21646 [0] NCCL INFO init.cc:815 -> 1
lo-s4-039:21587:21646 [0] NCCL INFO init.cc:951 -> 1
lo-s4-039:21587:21646 [0] NCCL INFO misc/group.cc:69 -> 1 [Async thread]

lo-s4-039:21614:21647 [1] transport/p2p.cc:574 NCCL WARN failed to open CUDA IPC handle : 711 peer mapping resources exhausted
lo-s4-039:21614:21647 [1] NCCL INFO init.cc:669 -> 1
lo-s4-039:21614:21647 [1] NCCL INFO init.cc:815 -> 1
lo-s4-039:21614:21647 [1] NCCL INFO init.cc:951 -> 1
lo-s4-039:21614:21647 [1] NCCL INFO misc/group.cc:69 -> 1 [Async thread]
Traceback (most recent call last):
  File "tools/lightning.py", line 514, in <module>
Traceback (most recent call last):
  File "/cluster/home/user/tracking/tools/lightning.py", line 514, in <module>
    trainer.fit(model)
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 451, in fit
    results = self.accelerator_backend.train()
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/pytorch_lightning/accelerators/ddp_backend.py", line 140, in train
    trainer.fit(model)
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 451, in fit
    results = self.ddp_train(process_idx=self.task_idx, model=model)
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/pytorch_lightning/accelerators/ddp_backend.py", line 266, in ddp_train
    model = model.configure_ddp(model, device_ids)
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/pytorch_lightning/core/lightning.py", line 954, in configure_ddp
    results = self.accelerator_backend.train()
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/pytorch_lightning/accelerators/ddp_backend.py", line 140, in train
    results = self.ddp_train(process_idx=self.task_idx, model=model)
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/pytorch_lightning/accelerators/ddp_backend.py", line 266, in ddp_train
    model, device_ids=device_ids, find_unused_parameters=True
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 333, in __init__
    model = model.configure_ddp(model, device_ids)
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/pytorch_lightning/core/lightning.py", line 954, in configure_ddp
    self.broadcast_bucket_size)
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 549, in _distributed_broadcast_coalesced
    model, device_ids=device_ids, find_unused_parameters=True
    dist._broadcast_coalesced(self.process_group, tensors, buffer_size)
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 333, in __init__
RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1595629403081/work/torch/lib/c10d/ProcessGroupNCCL.cpp:518, unhandled cuda error, NCCL version 2.4.8
    self.broadcast_bucket_size)
  File "/cluster/home/user/miniconda3/envs/track/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 549, in _distributed_broadcast_coalesced
    dist._broadcast_coalesced(self.process_group, tensors, buffer_size)
RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1595629403081/work/torch/lib/c10d/ProcessGroupNCCL.cpp:518, unhandled cuda error, NCCL version 2.4.8
```

### LINKED ISSUE #4171: DDP error after upgrading to v1.0.1
## 🐛 Bug

I'll start by saying that before I've upgrade to v1.0.1, I've used v0.9.0 with no apparent DDP issues.

After I've upgraded to v1.0.1, I'm having issues training with multiple GPUs using DDP backend.

**Initiating multi-gpu training in the following scenarios will result with an error:**

1. The first GPU (i.e. ID 0) is not included in the GPUs list. for example:
`python main.py --distributed_backend 'ddp' --gpus 1,2,3`

2. The GPUs list is not sequential. For example:
`python main.py --distributed_backend 'ddp' --gpus 0,2,3`

The above will result with the following error message:
`RuntimeError: cuda runtime error (101) : invalid device ordinal at /opt/conda/conda-bld/pytorch_1595629427478/work/torch/csrc/cuda/Module.cpp:59`

**Initiating multi-gpu training in the following scenarios will work as expected:**

`python main.py --distributed_backend 'ddp' --gpus 2`
or
`python main.py --distributed_backend 'ddp' --gpus 0,1,2`

### Environment

* CUDA:
        - GPU:
                - GeForce GTX 1080 Ti
                - GeForce GTX 1080 Ti
                - GeForce GTX 1080 Ti
                - GeForce GTX 1080 Ti
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.19.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.6.0
        - pytorch-lightning: 1.0.1
        - tqdm:              4.50.2
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                -
        - processor:         x86_64
        - python:            3.7.8
        - version:           #119-Ubuntu SMP Tue Sep 8 12:30:01 UTC 2020

### PR DESCRIPTION
**Pull request by awaelchli**
Fixes #3791 
Fixes #4171
Fixes #3865 (maybe, need more info)

All of these commands work now and run on the correct devices:

**running on gpu 1 and 2:**
```bash
python image_classifier.py --gpus "1,2" 
python image_classifier.py --gpus "1,2" --distributed_backend ddp
```

**running on gpu 1 and 2 with CUDA_VISIBLE_DEVICES:**
```bash
CUDA_VISIBLE_DEVICES="1,2" python image_classifier.py --gpus "0,1" 
CUDA_VISIBLE_DEVICES="1,2" python image_classifier.py --gpus "0,1" --distributed_backend ddp
# or equivalent
CUDA_VISIBLE_DEVICES="1,2" python image_classifier.py --gpus 2 --distributed_backend ddp
```

**non-consecutive ids also work**
running on gpu 1 and 3:
```bash
python image_classifier.py --gpus "1,3" 
python image_classifier.py --gpus "1,3"  --distributed_backend ddp
```

equivalent with CUDA_VISIBLE_DEVICES
```bash
CUDA_VISIBLE_DEVICES="1,3" python image_classifier.py --gpus "0,1" 
CUDA_VISIBLE_DEVICES="1,3" python image_classifier.py --gpus "0,1"  --distributed_backend ddp
# or 
CUDA_VISIBLE_DEVICES="1,3" python image_classifier.py --gpus 2  --distributed_backend ddp
```

### CODE REVIEW

**Review by teddykoker — APPROVED** [review]
LGTM :)

**Review by justusschock — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Comment by awaelchli** [comment]
please hold back with merge until I have @williamFalcon 's review/approval because it's his ddp code. Thanks

**Review by Borda — APPROVED** [review]

**Comment by williamFalcon** [comment]
looking now. this is great!

**Comment by awaelchli** [comment]
@ananthsub maybe you want to have a look here too. The more eyes on ddp the better :)

File: pytorch_lightning/accelerators/accelerator_connector.py
Comment (ananthsub) [inline pytorch_lightning/accelerators/accelerator_connector.py:209]: is there documentation around what env vars lightning sets anywhere?

File: pytorch_lightning/accelerators/ddp_accelerator.py
Comment (ananthsub) [inline pytorch_lightning/accelerators/ddp_accelerator.py:175]: do we need process_idx at all then?

**Review by ananthsub — APPROVED** [review]
I added a few n00b questions but LGTM!

File: pytorch_lightning/accelerators/accelerator_connector.py
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator_connector.py:209]: No, I am sure there is no documentation about these. They are internally used only, not meant to be modified by user. 
Should they be documented? I'm not sure

File: pytorch_lightning/accelerators/ddp_accelerator.py
Comment (awaelchli) [reply pytorch_lightning/accelerators/ddp_accelerator.py:175]: yes, we can probably remove it. the device the model is on is anyway not related to the process index, which was actually the root cause of the bug here.

File: pytorch_lightning/accelerators/accelerator_connector.py
Comment (ananthsub) [reply pytorch_lightning/accelerators/accelerator_connector.py:209]: ah I meant for contributors/developers primarily. I should probably read more of the code too :)

Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator_connector.py:209]: ah, good point. where could env variables be documented, they are like global variables. maybe at the top of file of the accelerator base class. I have no good suggestions :)

File: pytorch_lightning/accelerators/ddp_spawn_accelerator.py
Comment (ananthsub) [inline pytorch_lightning/accelerators/ddp_spawn_accelerator.py:165]: does this same fix apply to the ddp_torchelastic_accelerator? are there other DDP accelerators which need this change too?

Comment (awaelchli) [reply pytorch_lightning/accelerators/ddp_spawn_accelerator.py:165]: yes, some of these accelerators like torchelastic and ddp2 should also get this fix. The problem is, I didn't want to touch these because I cannot test the fix myself, lacking multi node setup. It would be good to follow up on this.

Comment (williamFalcon) [reply pytorch_lightning/accelerators/ddp_spawn_accelerator.py:165]: sorry, i actually don't think the others need this fix.

The reason is that torchelastic and co derive their process numbers from the environment, so they don't need to do this (local rank, global rank, etc).

Our DDP is a special case where we are acting like torchelastic, so we need to set this stuff up manually

**Comment by williamFalcon** [comment]
ok, just verified that this works!

**Comment by williamFalcon** [comment]
@ananthsub, just pushed an rc.

Mind checking on slurm and TE to make sure this is fine so we don't run into issues during the patch on Tuesday?

**Pull request merged**
